### PR TITLE
Add SEARCH_RESPONSE_TIME analytics event to searches

### PR DESCRIPTION
**Pull request by adjeiv**
## Fixes

Related to #2471 by @sarayourfriend . Currently only handles requests that are funnelled through the MediaService `search`, happy to extend/amend this if the current approach seems reasonable - just wanted to get some feedback beforehand.

## Description
This PR defines a new analytics event, the schema for which is described in #2471. A new SEARCH_RESPONSE_TIME event is sent for every search request that leads to a non-cached response, from which we can parse the headers to construct the event data. I've added some tests to handle various response cases.

One thing I wasn't completely certain about: do we want to source the queryString from the responseURL as is currently the case? And if so, do we want to include the leading question mark? Further, there's currently no way of telling whether this response came from e.g. an images or audio endpoint hit - wondering whether and how we'd want to incorporate that.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by sarayourfriend — COMMENTED** [review]
This looks really great to me, nice work @adjeiv. The unit tests are excellent :+1: 

I've left a comment suggesting how we can track the media type. My preference is to using separate events per media type because it allows more flexibility when we query the events. I'd also like your opinion on whether to include cache hit events as well. I am leaning towards yes, but if you have any thoughts regarding this or reasons we should exclude them beyond the original issue's description, please share them.

Really nice work :+1: Once the media type is tracked I will happily approve this PR.

File: frontend/src/data/media-service.ts
Comment (adjeiv) [reply frontend/src/data/media-service.ts:67]: All sounds good - I think having separate events makes sense to increase how usable the results are, so will get working on those changes. Just wanted to clarify on the second point - I was under the impression the Date header would help to clean out any local caching, rather than any Cloudflare caching. Do you mean that the local caching is something we'd also potentially like to track?

Comment (sarayourfriend) [reply frontend/src/data/media-service.ts:67]: Oh right! I completely forgot about that, you're right. The Cloudflare responses, even cached ones, will have dates not in the past.

I wrote the issue and the details of how it would work, and forgot this, so it would be good to clarify the comment to say "local cache" rather than just "cache". Our Cloudflare configuration caches some search responses, so the confusion will be easy to make for maintainers familiar with Openverse's infrastructure.

**Comment by adjeiv** [comment]
> This looks really great to me, nice work @adjeiv. The unit tests are excellent 👍
> 
> I've left a comment suggesting how we can track the media type. My preference is to using separate events per media type because it allows more flexibility when we query the events. I'd also like your opinion on whether to include cache hit events as well. I am leaning towards yes, but if you have any thoughts regarding this or reasons we should exclude them beyond the original issue's description, please share them.
> 
> Really nice work 👍 Once the media type is tracked I will happily approve this PR.

I've just made the changes per your comments.
On the issue itself you briefly mentioned we might want to apply this to `getRelatedMedia`, which would be easy to change, but would suffer from the fact we're only using a query string. As this queries by ID via the endpoint itself, there wouldn't be a query string at all (or just `peaks` in the case of audio) for getting related media. 

I'd also be happy to address this within the scope of this PR, perhaps we could include more (or the entirety) of the response url rather than just the query string. Let me know what your thoughts are.

**Review by sarayourfriend — APPROVED** [review]
Fantastic work, thanks very much @adjeiv :pray:

**Comment by sarayourfriend** [comment]
Oops, @adjeiv, I am seeing now the Playwright failures. Are you able to see if you can determine the cause? They do look related, because some of the failures are in the load more button, which uses the changed code paths.

File: frontend/src/data/media-service.ts
Comment (dhruvkb) [reply frontend/src/data/media-service.ts:71]: > it might be better to use separate event names per media type

A lot of current analytics events use a single event with a subfield for the media type (seen by searching usages of type `SearchType` in the file `frontend/src/types/analytics.ts`) so I think it is acceptable to do so here (or we might want to go back and update the existing ones?)

**Comment by AetherUnbound** [comment]
Hi @adjeiv, thanks so much for your contribution! We're currently in the midst of upgrading to a new version of Nuxt (see #3646). For the time being I'm going to draft this until we get that PR merged, then we can work on getting it applied. Thank you for your patience with us!

**Comment by AetherUnbound** [comment]
Undrafting this as we've unfrozen the frontend - I'll take a look at it today!

**Review by AetherUnbound — COMMENTED** [review]
This looks great! I tried running it locally though, and I'm not seeing any `*_SEARCH_RESPONSE_TIME` events showing up in the console logs when I make searches and navigate to media. Should those be showing up in the console? It makes me think they may not be getting recorded.

**Review by obulat — COMMENTED** [review]
The changes look awesome!

Since we are adding additional search views, that also perform search, we will need to add this event to those views as well. They would be different from the search events because they would have different paths instead of queries, at least for now. For example, the "Cat" tag view would be `/tag/cat`, and "Olga" creator view on Flickr would be `/source/flickr/creator/Olga`. Do you think adding a `path` property to the events is a good solution for this, @sarayourfriend?
In case we add some filtering to the additional search views in the future, they would also send the `query` parameter.

**Comment by sarayourfriend** [comment]
> Do you think adding a path property to the events is a good solution for this, @sarayourfriend?

No, because that would duplicate Plausible's built-in props, which are more flexible to query against: https://plausible.io/docs/stats-api#properties. It would duplicate `event:page` specifically.

> In case we add some filtering to the additional search views in the future, they would also send the query parameter.

We are sending the query string as `queryString` in the new event. Is that what you mean or are you referencing something else?

**Comment by obulat** [comment]
> > Do you think adding a path property to the events is a good solution for this, @sarayourfriend?
> 
> No, because that would duplicate Plausible's built-in props, which are more flexible to query against: https://plausible.io/docs/stats-api#properties. It would duplicate `event:page` specifically.
> 
> > In case we add some filtering to the additional search views in the future, they would also send the query parameter.
> 
> We are sending the query string as `queryString` in the new event. Is that what you mean or are you referencing something else?

I re-read the issue again, and I think agree that we don't need to add the `path` to the event. However, I think we could add `searchKind` or similar event that would allow us to distinguish the search events from the "collection" events (the search requests for additional search views). I don't think we need to go more granular and set the kind of the collection (tag/source/creator).

I also agree with Dhruv that we should not be splitting events based on the media type. Since we have other events that are not split on them, this will either require splitting all other events and increase the number of steps when adding a new media type, or will make the events set inconsistent. We should find a workaround in Plausible (or using Plausible data) instead. Do you know if that could be possible, @dhruvkb?

**Comment by sarayourfriend** [comment]
The issue is Plausible doesn't allow for querying against multiple custom props, which is a limitation of Clickhouse, the time series database Plausible uses, not really Plausible itself.

I suppose I was misled when I thought it made sense to divide search timing by media type, because actually the main thing we want to get here is the time it takes for people to receive search responses, and the actual response content between image and audio are not that different, with the largest amount of time usually being spent on the server generating the response, rather than network time sending the request and receiving the response, which is what this event would capture.

I think we might not need that granularity because of the lack of distinct difference between the two in response content _except_ that audio results return with the waveform on the frontend, which theoretically increases payload size. But if payload size is the question, we can record that instead.

We don't need to record search type because the collection pages all have distinct paths.

This really boils down to not being able to query against multiple custom props, and trying to leverage the built-in props as much as possible, because we can query against as many of those as we like... but only ever a single custom prop.

I don't have much more to add to this PR, and the review process on this has been so long and drawn out, that I am more eager to find a solution that will get two approvals rather than continue to advocate for any specific position I think is best.

So whomever is going to review this aside from me, please share the specific changes needed to get the approval so @adjeiv can update the PR, if they are still interested in working on this. If not, which is fine, especially with the considerable delay and several different changes requested, then someone else should take the PR over, make the changes, and get it re-reviewed.

**Review by dhruvkb — APPROVED** [review]
Personally I don't mind the per-media-type approach so I'll just approve it. We can in the future decide to combine them if the separate media type info isn't valuable or if it proves to be, we can also consider splitting up all the other events. For now this is a good PR so blocking it does not seem fair.

Thanks for the great contribution and for being so amenable to feedback @adjeiv.

**Pull request merged**
